CI: tests: reduce output by using chronic

grahamwhaley opened the pull request:
We only really want to see the tests that fail. Wrap the
test invocations in chronic to reduce the output.

Fixes: #809

Signed-off-by: Graham Whaley <graham.whaley@intel.com>

grahamwhaley commented:
Oh, and the only downside I can think of is that when we add new tests we won't be able to see if they actually ran via the logs, if they worked. We may have to verify they actually run by hand locally.

grahamwhaley commented:
Hmm, Travis error I don't understand. Could be travis wobble?
```
/home/travis/gopath/src/github.com/kata-containers/tests/stress
INFO: Installing xurls utility
package mvdan.cc/xurls/cmd/xurls: unrecognized import path "mvdan.cc/xurls/cmd/xurls" (https fetch: Get https://mvdan.cc/xurls/cmd/xurls?go-get=1: dial tcp 178.62.67.243:443: connect: connection refused)
The command "bash .ci/static-checks.sh github.com/kata-containers/tests" exited with 1.
Done. Your build exited with 1.
```

jodh-intel commented on Makefile:
```diff
@@ -39,55 +39,55 @@ functional: ginkgo
 	$(error RUNTIME is not set)
 else
-	./ginkgo -v functional/ -- -runtime=${RUNTIME} -timeout=${TIMEOUT}
+	chronic ./ginkgo -v functional/ -- -runtime=${RUNTIME} -timeout=${TIMEOUT}
```
How about doing something like this:

```
ifeq ($(V),)
    PRECMD=@chronic
endif

ifeq ($(V),1)
    PRECMD=@
endif

   :

$(PRECMD) ./ginkgo -v functional/ -- -runtime=${RUNTIME} -timeout=${TIMEOUT}
```

grahamwhaley replied:
Yep, that'd work. I did wonder if we might want some (default on) quiet mode. I'll see what the CIs think, and garner any further input - and then very likely make that change.

grahamwhaley commented:
CIs seem to be failing on [our code](https://github.com/kata-containers/tests/blob/master/.ci/static-checks.sh#L372-L378)

```bash
        local cmd="xurls"

	if [ ! "$(command -v $cmd)" ]
	then
		info "Installing $cmd utility"
		go get -u "mvdan.cc/xurls/cmd/$cmd"
	fi
```

fail is:
```
versions.yaml
  7:1       warning  missing document start "---"  (document-start)

INFO: Installing xurls utility
package mvdan.cc/xurls/cmd/xurls: unrecognized import path "mvdan.cc/xurls/cmd/xurls" (https fetch: Get https://mvdan.cc/xurls/cmd/xurls?go-get=1: dial tcp 178.62.67.243:443: connect: connection refused)
Build step 'Execute shell' marked build as failure
```

That domain does not respond for me - I wonder if we should try changing that to `https://github.com/mvdan/xurls/tree/master/cmd/xurls` ? /cc @chavafg and @jodh-intel as I believe you added that url 4 months back when we added the docs checks.

Looking at previous CI builds, it seems to have worked before.

I may spin a patch at the front of the series to try it out...

grahamwhaley commented on Makefile:
```diff
@@ -39,55 +39,55 @@ functional: ginkgo
 	$(error RUNTIME is not set)
 else
-	./ginkgo -v -focus "${FOCUS}" -skip "${SKIP}" ./integration/docker/ -- -runtime=${RUNTIME} -timeout=${TIMEOUT}
+	chronic ./ginkgo -v -focus "${FOCUS}" -skip "${SKIP}" ./integration/docker/ -- -runtime=${RUNTIME} -timeout=${TIMEOUT}
```
oh, yeah, you are probably right :-( mumble.... any thoughts on if there is anything else we can do or other options (`chronic` doesn't have a regular poll type output ...)

chavafg replied:
We used to have ginkgo without verbose, which suppress a lot of the messages we currently see in the log. But those are helpful when we need to debug why a test fail...

Taking a look at the [ginkgo docs](http://onsi.github.io/ginkgo/), I found these options which might help:
```
--succinct

Succinct silences much of Ginkgo’s more verbose output. Test suites that succeed basically get printed out on just one line! Succinct is turned off, by default, when running tests for one package. It is turned on by default when Ginkgo runs multiple test packages.

--trace

If present, Ginkgo will print out full stack traces for each failure, not just the line number at which the failure occurs.
```


Btw, the cri-o tests do not send much output, I think we can remove the chronic from those.

jodh-intel replied:
There is always the magic nul byte hack - if Travis/Jenkins are simply looking for any type of characters to avoid timing out, we could start a background process that writes a "`\0`" periodically. Something like one of:

- 
  ```
  $ utfout -b 1s '' -r -1
  ```

-
  ```
  $ while true; do printf "\0";sleep 1; done
  ```

grahamwhaley commented:
@jodh-intel - added a Makefile 'V' construct and wrapped it around `chronic`.
@chavafg - dropped the quieting of the `docker` ginkgo tests, and tweaked the arguments to see if that is less verbose, but still verbose on failure.

I'm having trouble testing this locally, as ginkgo doesn't seem to want to run on my machine. I therefore expect some feedback on this push, but really I want to see what happens in the CIs (does it run at all, and if so, what does it look like...)
Feedback most welcome..

grahamwhaley commented:
/test

grahamwhaley commented:
Looks like the crio tests time out as well:

```
Ginkgo ran 1 suite in 18m41.55925028s
Test Suite Passed
     TEST    docker
     TEST    crio
     TEST    crio
Build timed out (after 5 minutes). Marking the build as aborted.
```

I'll tweak and re-spin...

grahamwhaley commented:
/test

grahamwhaley commented:
Hmm, and cannot ignore the k8s tests either. /me wonders if we might want to bump that 5minute silence timeout @chavafg , but for now, I'll go and see what I can do for the k8s tests...

```
sys	0m27.948s
/tmp/jenkins/workspace/kata-containers-tests-ubuntu-16-04-PR/go/src/github.com/kata-containers/tests
     TEST    docker-compose
     TEST    docker-compose
     TEST    docker-stability
     TEST    docker-stability
     TEST    openshift
     TEST    openshift
     TEST    kubernetes
     TEST    kubernetes
Build timed out (after 5 minutes). Marking the build as aborted.
Build was aborted
```

grahamwhaley commented:
/test

grahamwhaley commented:
CIs seem happy! And a 16.04 CI log is a 'mere' 367k. Let me see if I can find a previous one to compare to see if we did actually make any difference.....
Well, the previous pass comes in at 1179k and one before that at 1210k - so, maybe we did win!

marcov commented on Makefile:
```diff
@@ -39,57 +43,60 @@ functional: ginkgo
-	systemctl is-active --quiet docker || sudo systemctl start docker
-	cd integration/stability && \
-	export ITERATIONS=2 && export MAX_CONTAINERS=20 && chronic ./soak_parallel_rm.sh
+	$(QUIET_TEST) systemctl is-active --quiet docker || sudo systemctl start docker
```
for really muting this, you'd need `$(QUIET_TEST)` on both sides of `||`. But I am fine with how it is now :)

marcov commented:
@grahamwhaley isn't it worth muting all the logs by finding a solution to the timeout issue?
You could create a wrapper script around chronic and implement @jodh-intel idea like this:

```bash
cmdLine=$@

chronic ${cmdLine[@]} &
cmdPid="$!"

while true; do printf "\0";sleep 1; done
printerPid="$!"

wait "$cmdPid"
kill "$printerPid"
```

sboeuf commented:
@grahamwhaley this looks good to me. Just let us know if you're planning to rework as suggested by @marcov and @jodh-intel, otherwise we can simply go ahead and merge it.

grahamwhaley commented:
OK @marcov @jodh-intel - I've added a little `chronic` wrapper to the .ci dir and invoked that from the Makefile instead of `chronic` itself. Let's see if it flies... or if I need to munge a bit more.

chavafg commented:
/test

grahamwhaley commented:
oops - we forgot to wrap the sub-sleep-shell in `()&`, so that attempt hung up forever - edit, push, try again and....
/test

marcov commented:
> oops - we forgot to wrap the sub-sleep-shell in `()&`, so that attempt hung up forever - edit, push, try again and....

yep, we want our tests to finish eventually :smile:

lgtm

sboeuf commented:
/test

jodh-intel commented:
lgtm

/test

raravena80 commented:
@grahamwhaley ping (from your weekly Kata herder)

grahamwhaley commented:
I'm seeing a test timeout on the ubuntu fail at least:
```
     TEST    kubernetes
Build timed out (after 5 minutes). Marking the build as aborted.
Build was aborted#
```

I wonder if that is sporadic, specific to some setups, or what.
I might modify the 'NULL echo' into an echo of a `.` maybe, which might make it clearer it is running, and maybe help with the timeouts. Let me nudge the CIs again right now though and see what happens...

grahamwhaley commented:
re-triggered the failed CIs - if they fail again I'll dig deeper...

marcov commented:
> I'm seeing a test timeout on the ubuntu fail at least:
> 
> ```
>      TEST    kubernetes
> Build timed out (after 5 minutes). Marking the build as aborted.
> Build was aborted#
> ```
> I wonder if that is sporadic, specific to some setups, or what.
> I might modify the 'NULL echo' into an echo of a `.` maybe, which might make it clearer it is running, and maybe help with the timeouts. 

Same again, I'd try the `echo -n "."` at this point.

grahamwhaley commented:
Updated to use '.'s for the progress (instead of NULLs). Also rebased to master and fixed up the new (and some not-quite-perfect) rules to quieten them. let's see how it flys and looks in the CIs.
We can probably start with a peek at travis before we kick a full test cycle.

jodh-intel commented:
:cry: But I prefer dashes to dots...

Kiddin'

lgtm

grahamwhaley commented:
heh, if I'd had the bash foo energy then you might have gotten...
```
----5----10----15---- etc.
```
:-)
/test

jodh-intel commented:
Could I also have :star: :star2: and :unicorn:'s? :smile:

jodh-intel commented:
Oh, and :rainbow:'s of course!

jodh-intel commented:
/me seems an opportunity for the Unicode Consortium to create `:rainbow_left:` and `:rainbow_right`.

grahamwhaley commented:
hurumph... errr, so, now what do we do @jodh-intel @marcov ...
I think I'm going to have to put a carriage return on the end of the `.`s (or every now and then....). Let me construct a slightly more complex sleeper loop - @jodh-intel - you may get your wish for more characters beyond `.`...
```
.
     TEST    kubernetes
.
     TEST    kubernetes
............................................................Build timed out (after 5 minutes). Marking the build as aborted.
Build was aborted
Performing Post build task...
```

marcov commented:
Now I am really curious to see what the jenkins code for the timeout is doing.

jodh-intel commented:
```
... --- ...
```

grahamwhaley commented:
repushed.
I'm sure the output is nowhere near as humorous as @jodh-intel might have liked... but, hey, you can PR over it ;-)
/test

grahamwhaley commented:
/test

grahamwhaley commented:
For fun, I wanted to see what it looked like in the logs....

```
     TEST    functional
.
     TEST    functional
.
     TEST    docker
.10.20.30.
     TEST    docker
.
     TEST    crio
.
     TEST    crio
.
     TEST    docker-compose
.
     TEST    docker-compose
.
     TEST    network
.
.
     TEST    network
.
     TEST    network
.10.
     TEST    network
.10.20.
     TEST    netmon
.
.
     TEST    netmon
.
     TEST    netmon
.
     TEST    docker-stability
.
.
     TEST    docker-stability
.
     TEST    openshift
.
     TEST    openshift
.
     TEST    kubernetes
.
     TEST    kubernetes
.10.20.30.40.50.60:
.10.20.30.40.50.60:
.10.20.30.40.50.60:
.10.20.30.40.50.60:
.10.20.30.40.50.60:
.10.20.30.40.50.60:
.
     TEST    swarm
.
.
     TEST    swarm
.
     TEST    swarm
.
     TEST    vm-factory
.
     TEST    entropy
.
     TEST    entropy
.Command 'cd' not found in /tmp/jenkins/workspace/kata-containers-tests-ubuntu-16-04-PR/go/bin, /usr/local/go/bin, /usr/sbin, /sbin, /usr/local/sbin, /usr/local/bin, /usr/sbin, /usr/bin, /sbin, /bin, /usr/games, /usr/local/games at /usr/bin/chronic line 44.

/bin/sh: 2: @echo: not found
.bats: /tmp/jenkins/workspace/kata-containers-tests-ubuntu-16-04-PR/go/src/github.com/kata-containers/tests/entropy_test.bats does not exist

     TEST    ramdisk
.10.
```

So, there are a couple of interesting things in there - like, at the end, the `cd` and `echo` items for what looks like the entropy test. Is that something for you @GabyCT ?
That'll also look like 6 minutes to run the k8s tests.

Other than that, it looks 'useable'. Now, one day I may attack the what I think is kernel build output in the logs, which I think come from the osbuilder maybe - and then maybe look at the monster logs we get on failure (as mostly I find I have to scroll 22Mb of logs backwards over the log output to find the error message - I'd like to  see the logs as attachements (only) :-) )

jodh-intel commented:
Just a thought ...

`printf "\0"` didn't work and now we have dots. But what about the following which should avoid polluting the test output:

```
printf ".\b"
```

grahamwhaley commented:
I was going to nudge for a merge here, but peeking at the logs (looking for something else), I see:

```
     TEST    netmon
.10.
     TEST    docker-stability
.
.
     TEST    docker-stability
.
     TEST    openshift
.
```

which I find suspicious, as I expect the `docker-stability` tests to take longer than that, as they include the soak test... hmm. Here is the snippet of patch change - maybe that is not doing what I thought...

```patch
docker-stability:
-	systemctl is-active --quiet docker || sudo systemctl start docker
-	cd integration/stability && \
-	export ITERATIONS=2 && export MAX_CONTAINERS=20 && chronic ./soak_parallel_rm.sh
+	$(QUIET_TEST) systemctl is-active --quiet docker || \
+	$(QUIET_TEST) sudo systemctl start docker
+	$(QUIET_TEST) bash -c "export ITERATIONS=2; export MAX_CONTAINERS=20; ./integration/stability/soak_parallel_rm.sh"
```

I'd better mark as DNM, and check that....

grahamwhaley commented:
> ```
> printf ".\b"
> ```
That's a backspace, yes - hmm, that may still turn up in the logs as expanded magic codes (dunno) - and I quite like having the timings printed out at the minute, so we can have some sort of reference in the logs to how long each test stage took.

Sure, if I need to update the PR for the stability niggle then maybe I'll give it a try. Not sure how I'll handle the periodic time updates (10..20) - would probably have to drop them (as I don't want to be calculating how many backspaces to print ;-))

grahamwhaley commented:
@GabyCT - oh, I know what the `@echo not found` issue is - not you ;-) - it is using the `$(QUIET_TEST)` in a compound line that makes bash then see the `@echo` makefile construct, and bawk. grrr. @jodh-intel @marcov - this just gets more and more complex :-)

grahamwhaley commented:
OK, @jodh-intel @marcov - going to call on your Makfile and bash-foo here for some input. Any insights on how we make this work when converting a line such as:

```bash
	cd integration/stability && \
	export ITERATIONS=2 && export MAX_CONTAINERS=20 && chronic ./soak_parallel_rm.sh
```

which currently I turned into:

```bash
	$(QUIET_TEST) bash -c "export ITERATIONS=2; export MAX_CONTAINERS=20; ./integration/stability/soak_parallel_rm.sh"
```

which does not work, as the `"`s around the bash command don't make it through to the `kata_chronic.sh` script to then handle the chronic and liveness prints.

I can't bury the call to `$(QUIET_TEST)` inside lines either, as then the `@echo` expansion gets interpreted as a shell item, and balks....

I'm running out of ideas...

jodh-intel commented on .ci/kata_chronic.sh:
```diff
@@ -0,0 +1,32 @@
+
+cmdLine=$@
+
+chronic ${cmdLine[@]} &
```
@grahamwhaley - I think the fix you need is to use `eval`:

```
eval chronic ${cmdLine[@]} &
```

grahamwhaley replied:
Thanks @jodh-intel - that seems to have done the trick. I re-massaged all the make lines with `&&` or `||` requirements into `bash -c` lines to get rid of the `@echo` faux-pas as well.
I didn't do the `printf ".\b"` update though - as I think I'd have to invest time to make it 'nice' in the log output and keep some semblance of printing 'how long did it actually take'. We can PR updates to that later if we find we need.
/test

grahamwhaley commented:
/test

jodh-intel commented:
lgtm

note - still has the dnm label.

grahamwhaley commented:
Well, I think it is working better now, but there is still a bunch of output I didn't think we'd see:

```
     TEST    docker-compose
.
     TEST    docker-compose
.bats: /tmp/jenkins/workspace/kata-containers-tests-ubuntu-16-04-PR/go/src/github.com/kata-containers/tests/docker-compose.bats does not exist

     TEST    network
.
     TEST    network
.
     TEST    network
.1..1
not ok 1 ping container with macvlan driver
# (from function `clean_env' in file integration/network/macvlan/../../../lib/common.bash, line 83,
#  from function `setup' in test file integration/network/macvlan/macvlan_driver.bats, line 29)
#   `clean_env' failed
# Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
# Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
# Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
# 

     TEST    network
.10.1..2
not ok 1 ping container with ipvlan driver with mode l2
# (from function `clean_env' in file integration/network/ipvlan/../../../lib/common.bash, line 83,
#  from function `setup' in test file integration/network/ipvlan/ipvlan_driver.bats, line 26)
#   `clean_env' failed
# Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
# Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
# "docker network rm" requires at least 1 argument.
# See 'docker network rm --help'.
# 
# Usage:  docker network rm NETWORK [NETWORK...]
# 
# Remove one or more networks
# sed: no input files
# 
ok 2 ping container with ipvlan driver with mode l3

     TEST    netmon
.
     TEST    netmon
.
     TEST    netmon
.
     TEST    docker-stability
.
     TEST    docker-stability
.Enabling netmon at the configuration file
Checking Kata runtime kata-runtime
1.14: Pulling from library/nginx
a5a6f2f73cd8: Already exists
2343eb083a4e: Pulling fs layer
251439d5b33c: Pulling fs layer
251439d5b33c: Verifying Checksum
251439d5b33c: Download complete
2343eb083a4e: Verifying Checksum
2343eb083a4e: Download complete
2343eb083a4e: Pull complete
251439d5b33c: Pull complete
Digest: sha256:32fdf92b4e986e109e4db0865758020cb0c3b70d6ba80d02fe87bad5cc3dc228
Status: Downloaded newer image for nginx:1.14
Start iteration 1 of 5
Running...
```

I'll have a peek, but maybe that is for another day, so dropping the DNM now...

chavafg commented:
I am not sure this is working the desired way. I took at the ubuntu job and see some failures and still reported as success:

```
not ok 1 ping container with macvlan driver
# (from function `clean_env' in file integration/network/macvlan/../../../lib/common.bash, line 83,
#  from function `setup' in test file integration/network/macvlan/macvlan_driver.bats, line 29)
#   `clean_env' failed
# Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
# Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
# Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
# 

     TEST    network
.10.1..2
not ok 1 ping container with ipvlan driver with mode l2
# (from function `clean_env' in file integration/network/ipvlan/../../../lib/common.bash, line 83,
#  from function `setup' in test file integration/network/ipvlan/ipvlan_driver.bats, line 26)
#   `clean_env' failed
# Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
# Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
# "docker network rm" requires at least 1 argument.
# See 'docker network rm --help'.
# 
# Usage:  docker network rm NETWORK [NETWORK...]
# 
# Remove one or more networks
# sed: no input files
# 
```
http://jenkins.katacontainers.io/job/kata-containers-tests-ubuntu-16-04-PR/838/console

jodh-intel commented on .ci/kata_chronic.sh:
```diff
@@ -0,0 +1,32 @@
+
+printerPid="$!"
+
+wait "$cmdPid"
```
Ah - @chavafg is right. This needs to be something like

```sh
wait "$cmdPid"
ret=$?

printf "\n"
kill "$printerPid"

# Return the exit code of the command run, not of the `kill`.
exit "$ret"
```

grahamwhaley replied:
cool - thanks for the spot - it indeed didn't quite feel right...

jodh-intel commented:
Whilst you're at it, could you add the new script to the CI tests readme?:

- https://github.com/kata-containers/tests/blob/master/.ci/README.md

grahamwhaley commented:
@chavafg I'll see if I can find a nice way to only do it when CI=true (any hints for how to do that in the Makefile welcome.. beyond an if/else)
@jodh-intel sure, I'll go do docs as well.
DNM re-applied.
Who would have thought adding `chronic` could take >50 github comments eh?

grahamwhaley commented:
done:
- added an if/else to cater for CI=true (I probably could have done it with just more magic makefile variable substitutions, but I didn't really want to add more obfuscation to that magic...)
- added chronic (I really should have called that kronic ;-)) to the readme
- fixed the exit value for the chronic script

grahamwhaley commented:
/test

jodh-intel commented on .ci/README.md:
```diff
@@ -20,6 +20,7 @@ This directory contains scripts used by the [Kata Containers](https://github.com
 | `install_*` | Install various parts of the system and dependencies. |
 | `jenkins_job_build.sh` | Called by the [Jenkins CI](https://github.com/kata-containers/ci) to trigger a CI run. |
 | `kata-arch.sh` | Displays architecture name in various formats. |
+| `kata_chronic.sh` | Runs commands under `chronic`, whilst generating a heatbeat ticker. |
```
Nit: typo "heartbeat".

chavafg commented:
Thanks @grahamwhaley for adding the `if/else` for `CI` envar.
I now see this issue:
```
.bats: /tmp/jenkins/workspace/kata-containers-tests-vsock-fedora-27-PR/go/src/github.com/kata-containers/tests/docker-compose.bats does not exist
```

Not sure why it is not going inside `integration/docker-compose`.

chavafg commented on Makefile:
```diff
@@ -39,77 +47,73 @@ functional: ginkgo
-	cd integration/docker-compose && \
-	bats docker-compose.bats
+	$(QUIET_TEST) bash .ci/install_bats.sh
+	$(QUIET_TEST) bash -c "cd integration/docker-compose && bats docker-compose.bats"
```
what about: 
```
bash -c "bats integration/docker-compose/docker-compose.bats"
```

grahamwhaley replied:
Oh, this is really frustrating.... can I scream yet? :-/
ouch ouch ouch.... so, it seems the `&&`'s in the Makefile, when they hit the eval, are separated out as a separate (non-eval'd) command. and hence, it fails. You can see it if you add a `set -x` to the kata_chronic.sh script. Oh, and the Makefile still works fine with `CI=false`. grrrr.

All uses of `&&` in the Makefile were for `bats`, and luckily it looks like converting them to a direct `bats the/full/path` works, so I've done that.
Of course, then I was concerned that all use of `||` and `;` in the Makefile would have a similar problem. I think I've managed to convert the use of `;`, and the use of `||` I think will work (although maybe not be so quiet), as the return value from the initial `QUIET` call will still be valid.
sigh. OK, let's spin the wheel again....

grahamwhaley commented:
/test

jodh-intel commented on .ci/kata_chronic.sh:
```diff
@@ -0,0 +1,36 @@
+
+cmdLine=$@
+
+eval chronic ${cmdLine[@]} &
```
This is prolly safer if quoted, which might also help the issue you were seeing with `&&` / `||`:

```suggestion
eval chronic "${cmdLine[@]}" &
```

grahamwhaley replied:
tried it already - didn't work :-(  I even put the `"`s outside the `chronic` as well to try and wrap it further for the eval - still didn't work.... but, sure, quoting it is probably the right thing to do (but it doesn't fix the `&&` and `||` issue afaict). I'll edit and push

grahamwhaley commented:
/test

grahamwhaley commented:
Oh, I probably have a typo somewhere, or the bats files are making a path assumption - I'll have a look on Monday:
```
     TEST    entropy
.bats: /tmp/jenkins/workspace/kata-containers-tests-ubuntu-16-04-PR/go/src/github.com/kata-containers/tests/integration/entropy/bats does not exist
/usr/libexec/bats-exec-suite: line 20: let: count+=: syntax error: operand expected (error token is "+=")

Makefile:111: recipe for target 'entropy' failed
```

jodh-intel commented:
Hi @grahamwhaley - any update? There is now a conflict on this PR and I'm going to add a festive :chestnut: for giggles.

This has to win the prize for "*PR that sounded like a 5 minute task but which turned into a multi-headed killer razor-toothed tentacular swamp monster*" award.

jodh-intel commented:
Seriously? No `:monster:`? I think I may just have identified another opportunity for the Unicode Consortium to work on next year! :smile:

grahamwhaley commented:
oops, fixed typos and a leftover commented out line in the Makefile.... pushed
/test
(it also occurs to me that we could probably clean up the `bats` install calls into a single rule and tie many of the other rules to it - but, that's for another day...

jodh-intel commented:
Looks like we still have a Makefile conflict?

grahamwhaley commented:
rebased - try again...
/test

chavafg commented:
I was taking a look at the job times and it is taking 14-16 minutes with this changes instead of the 1hr, so seems like something is not going well...

grahamwhaley commented:
good call @chavafg - yeah, it is hard to trust it when you don't see anything...
I think the only/best way to debug it is to run with a `set -x` inside the chronic script, and then we can see what is really happening. I'll try to do that locally, and also time it with `V=1` on the Make..

grahamwhaley commented:
OK, time to try again - the CI logs have expired, so I cannot go look at where we got to - and, I wanted to add a simple bit of `echo` debug in.... repushed... let's see what the CIs make of it.
And, maybe I'll have to rebase and try again tomorrow...
/test

grahamwhaley commented:
@chavafg, @marcov @jodh-intel  - looking at the logs, I'm going to see if I can figure out why the stability test here looks like it is not taking ~4 minutes to run, like we expect it to:

```
     TEST    docker-stability
TESTING [bash -c ITERATIONS=2 MAX_CONTAINERS=20 ./integration/stability/soak_parallel_rm.sh
.
systemctl is-active --quiet docker || sudo systemctl start docker
cd integration/oci_calls && \
bash -f oci_call_test.sh
Running setup
Check oci calls for run
db5bb71509f16ff0c75d5909bb84f24b11949a3ddf59b7bd19d4093283cd154a
Check oci calls for stop
testoci
testoci
Check oci calls for run with true
Teardown
     TEST    openshift
TESTING [bash -f .ci/install_bats.sh
.
```

grahamwhaley commented:
I'm thinking of giving up on this PR as a lost cause, unless somebody can come up with a smarter more 'bash and make friendly' way of doing this.
The big problem generally stems from when we are trying to invoke tests that require some bash env setup etc. Take for example the `docker-stability` line (as that is easy to understand and run locally).

It currently looks like:
```bash
        $(QUIET_TEST) bash -c "ITERATIONS=2 MAX_CONTAINERS=20 ./integration/stability/soak_parallel_rm.sh"
```

which works for the general `make docker-stability` case, but if you try that with the `CI` enabled, what happens afaict is when we pass everything through to the `eval` in the chronic script, we drop the `"`s, and everything falls apart:

```
 CI=true make docker-stability
     TEST    docker-stability
TESTING [bash -c systemctl is-active --quiet docker || sudo systemctl start docker
.
     TEST    docker-stability
TESTING [bash -c ITERATIONS=2 MAX_CONTAINERS=20 ./integration/stability/soak_parallel_rm.sh
.
```

That finishes on my machine immediately, and passes - which is wrong, as on my local machine the test actually fails as I don't have the `ksm-throttler` running... let's tweak that and run it by hand to see what actually happens then - stick a `-x` on the command line and...:

```
$ bash -xc ITERATIONS=2 MAX_CONTAINERS=20 ./integration/stability/soak_parallel_rm.sh
+ ITERATIONS=2
```

Oh, it only ran the first bit. One might think we could solve that by putting `;`s or `&&`s between the sequence, like we had in the original Makefile, but if we do that then the eval ends up seeing them the `chronic` only binds to the first one, and we still get all the output from the actual tests....

The root problem is I cannot see a way to pass any wrapping `"`s into the script to then contain the command to bash, nor can I come up with another way of making the dependant commands a sequence that will bind together through the `eval` and thus get wrapped by a single `chronic`. Unless somebody can come up with an alternative or some bash/make foo-magic (oh, apparently known as [foo foo dust](https://www.urbandictionary.com/define.php?term=foo%20foo%20dust) :-) ), then I give up....
/cc @jodh-intel @marcov @chavafg for extra foo-ness....

jodh-intel commented:
One solution would be to mandate that:

- All the shell logic in `Makefile` is moved to a script.
- All scripts are made executable (so no need for the `bash -c` cruft).
- If a script needs config passed to it, it must be done via a CLI option.

That way `Makefile` would do something like:

```make
$(QUIET_TEST) ./integration/stability/soak_parallel_rm.sh --iterations $(ITERATIONS) --containers $(MAX_CONTAINERS)
```

... and `soak_parallel_rm.sh` would do the checking itself:

```sh
systemctl is-active --quiet docker || sudo systemctl start docker
```

ydjainopensource commented:
@grahamwhaley Are you working on this one? I will just add a don't merge and wip to separate this one out.

/do-not-merge
/wip

jodh-intel commented:
@grahamwhaley - I've added a dnm label. Do you plan to continue working on this, or are we going to give up?

jodh-intel commented:
@grahamwhaley - what's you feeling on this? Do we proceed with the plan (mentioned here: https://github.com/kata-containers/tests/pull/810#issuecomment-446947525), or close?

caoruidong commented:
ping @grahamwhaley

grahamwhaley commented:
Much though I'd like to have figured out this `chronic` 'quieten the logs', and the related #994 , the CI 'timeout on silence' just made it too hard right now. I'll close both these PRs for now, as I'm pretty sure I won't get the time to invest the effort needed to find a nice workable solution (which may involve having our own `chronic` or adding a `chronic` heartbeat option upstream...).
